accesscontrol/jwk.go

```diff
@@ -0,0 +1,173 @@
+package accesscontrol
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"reflect"
+)
+
+type JWK struct {
+	Key       interface{}
+	KeyID     string
+	Algorithm string
+	Use       string
+}
+
+type rawJWK struct {
+	Alg string `json:"alg"`
+	Kid string `json:"kid"`
+	Kty string `json:"kty"`
+	Use string `json:"use"`
+	// RSA public key
+	E   *base64URLEncodedField `json:"e"`
+	N   *base64URLEncodedField `json:"n"`
+	X5c []*base64EncodedField  `json:"x5c"`
+	//X5t *base64URLEncodedField `json:"x5t"`
+	//X5tS256" *base64URLEncodedField `json:"x5t#S256"`
+}
+
+func (j JWK) MarshalJSON() ([]byte, error) {
+	var raw *rawJWK
+	switch key := j.Key.(type) {
+	case *rsa.PublicKey:
+		raw = fromRsaPublicKey(key)
+	default:
+		return nil, fmt.Errorf("kty '%s' not supported", reflect.TypeOf(key))
+	}
+	raw.Kid = j.KeyID
+	raw.Alg = j.Algorithm
+	raw.Use = j.Use
+
+	return json.Marshal(raw)
+}
+
+func (self *JWK) UnmarshalJSON(data []byte) error {
+	var raw rawJWK
+	err := json.Unmarshal(data, &raw)
+	if err != nil {
+		return err
+	}
+
+	var key interface{}
+
+	switch raw.Kty {
+	case "RSA":
+		if len(raw.X5c) > 0 {
+			certificate, err := x509.ParseCertificate(raw.X5c[0].data)
+			if err != nil {
+				return fmt.Errorf("Invalid x5c")
+			}
+
+			key = certificate.PublicKey.(*rsa.PublicKey)
+		} else if raw.N != nil && raw.E != nil {
+			key = &rsa.PublicKey{
+				N: raw.N.toBigInt(),
+				E: raw.E.toInt(),
+			}
+		} else {
+			fmt.Printf("Ignoring invalid %s key: %q", raw.Kty, raw.Kid)
```

Comment:
`fmt.Errorf()`?

Reply:
`Errorf` will be rejected by static analyzer.